crates/libcontainer/src/network/network_device.rs

```diff
@@ -0,0 +1,263 @@
+use std::fs::File;
+use std::os::fd::AsRawFd;
+
+use netlink_packet_route::address::{AddressAttribute, AddressFlags, AddressMessage, AddressScope};
+use oci_spec::runtime::LinuxNetDevice;
+
+use super::address::AddressClient;
+use super::link::LinkClient;
+use super::wrapper::create_network_client;
+use super::Result;
+use crate::network::serialize::SerializableAddress;
+
+/// dev_change_netns allows to move a device given by name to a network namespace given by nsPath
+/// and optionally change the device name.
+/// The device name will be kept the same if device.Name is None or an empty string.
+/// This function ensures that the move and rename operations occur atomically.
+/// It preserves existing interface attributes, including IP addresses.
+pub fn dev_change_net_namespace(
+    name: String,
+    netns_path: String,
+    device: &LinuxNetDevice,
+) -> Result<Vec<SerializableAddress>> {
+    tracing::debug!(
+        "attaching network device {} to network namespace {}",
+        name,
+        netns_path
+    );
+
+    let mut link_client = LinkClient::new(create_network_client())?;
+    let mut addr_client = AddressClient::new(create_network_client())?;
+
+    let netns_file = File::open(netns_path)?;
+
+    let new_name = device
+        .name()
+        .as_ref()
+        .filter(|d| !d.is_empty())
+        .map_or(name.clone(), |d| d.to_string());
+
+    let link = link_client.get_by_name(&name)?;
+
+    let index = link.header.index;
+
+    // Set the interface link state to DOWN before modifying attributes like namespace or name.
+    // This prevents potential conflicts or disruptions on the host network during the transition,
+    // particularly if other host components depend on this specific interface or its properties.
+    link_client.set_down(index)?;
+
+    // Get the existing IP addresses on the interface.
+    let addrs = addr_client.get_by_index(index)?;
+
+    link_client
+        .set_ns_fd(index, &new_name, netns_file.as_raw_fd())
+        .map_err(|err| {
+            tracing::error!(?err, "failed to set_ns_fd");
+            err
+        })?;
+
+    let serialize_addrs: Vec<SerializableAddress> =
+        addrs.iter().map(SerializableAddress::from).collect();
+
+    Ok(serialize_addrs)
+}
+
+/// setup_network_device sets up a network device in a new namespace.
+/// It moves the device to the new namespace and adds the IP addresses to the device.
+/// It also sets the device up.
+pub fn setup_network_device(
```

Comment:
This function has two roles.
1. Assign an address.
2. Link up.

Can these parts be separated into two functions?